codecs/src/main/scala/com/jpmc/codecs/Encoder.scala

```diff
@@ -0,0 +1,73 @@
+package com.jpmc.codecs
+
+import com.jpmc.json._
+
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+trait Encoder[A] {
+  def encode(value: A): Json
+
+  def contraMap[B](f: B => A): Encoder[B] = Encoder.from { b => this.encode(f(b)) }
+}
+
+object Encoder {
+  def apply[A](implicit encoder: Encoder[A]): Encoder[A] = encoder
+
+  def from[A](f: A => Json): Encoder[A] = (value: A) => f(value)
+
+  implicit class EncoderSyntax[A](value: A) {
```

Comment:
Why do you have to define the asJson function in this class like this?

Reply:
thanks for the review @monmcguigan 

This pattern is known as "extension methods" or "pimp my class" (analogy to MTVs "pimp my ride" show)
It enables one to define a method on a type, and use it as if the type had the method already.
Consider the `Int` class, one can define a method to be used on integers! 
```scala

// Int is the type to which new methods are being added
implicit class PimpMyInteger(value: Int) {    
   def beefUp(by: Int) {
      value * by
   }
   
   // can add any number of methods
   def toneDown(by: Int) {
     value / by 
   }
}
```

now this new method can be used as a nifty way to multiple two numbers or divide them

```scala

val beefedUp = 2.beefUp(3)
// OR
val beefedUp = 2 beefUp 3

// take it down a notch
val tonedDown = 100 toneDown 4
```
`beefUp` is something which scala language does provide give on the integer type, but we made it look like this method came from integer type in scala language.

In scala 2 this involves creating an `implicit class`
This pattern has been syntactically formalised in [scala 3 with `extension` keyword](https://docs.scala-lang.org/scala3/reference/contextual/extension-methods.html)
Another language which has this feature is [Kotlin](https://kotlinlang.org/docs/extensions.html) 

With the `EncoderSyntax` class that is defined here, it is in an abstraction level above - the type on which the extension method is defined itself is a parameter `EncoderSyntax[A]`, and `asJson` will work as long as the type on which `asJson` is being called has an implicit instance of the Encoder in the scope!

Reply:
ahhhhh cool!! Thanks for this explanation, that makes complete sense.